src/core/models/dt_linearmodel.h

```diff
@@ -0,0 +1,181 @@
+//------------------------------------------------------------------------------
+// Copyright 2021 H2O.ai
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+// IN THE SOFTWARE.
+//------------------------------------------------------------------------------
+#ifndef dt_MODELS_LINEARMODEL_h
+#define dt_MODELS_LINEARMODEL_h
+#include "models/utils.h"
+#include "models/dt_linearmodel_base.h"
+#include "str/py_str.h"
+#include "models/label_encode.h"
+
+
+namespace dt {
+
+
+/**
+ *  Class template that implements all the virtual methods declared in
+ *  dt::LinearModelBase.
+ */
+template <typename T /* float or double */>
+class LinearModel : public dt::LinearModelBase {
+  private:
+    // SType that corresponds to `T`
+    const SType stype;
+    size_t : 56;
+    // Model's shape is (nfeatures + 1, nlabels)
+    dtptr dt_model;
+    std::vector<T*> w;
+    LinearModelType model_type;
+
+    // Feature importances datatable of shape (nfeatures, 2),
+    // where the first column contains feature names and the second one
+    // feature importance values.
+    dtptr dt_fi;
+
+    // Linear model parameters provided to constructor.
+    LinearModelParams params;
+
+    // Individual parameters converted to T type.
+    T eta;
+    T lambda1;
+    T lambda2;
+    T nepochs;
+
+    // Vector of feature interactions.
+    std::vector<sztvec> interactions;
```

Comment:
Yes, could be like that also. I believe that when I first added this parameter to the model (FTRL at that time), we didn't have virtual columns, so having `interactions` parameter could save a lot from the memory perspective and also from the hashing perspective. In this linear model I kept it because it seems pretty convenient for the user to simply pass the interactions as a list of lists, instead of adjusting the frame.

Reply:
If the model doesn't need to know that these are special "interactions" columns (and it usually doesn't), then a cleaner solution would be to separate these 2 concerns: perhaps create a separate function that augments the frame with interactions, and then let the user call that function if they want to.
 After all, creating interactions could use many different settings, and it just makes no sense to put all that complexity into the model itself.

Reply:
TODO: leave interactions creation to the user, also check FTRL